cli/local/model_cache.go

```diff
@@ -18,146 +18,123 @@ package local
-	modelPaths := make([]string, len(apiSpec.Predictor.Models))
-	for i, modelResource := range apiSpec.Predictor.Models {
-		modelPaths[i] = modelResource.ModelPath
-	}
+func CacheLocalModels(apiSpec *spec.API, onlyLocalModels bool) error {
+	var err error
+	var wasAlreadyCached bool
+	var localModelCache *spec.LocalModelCache
+	localModelCaches := make([]*spec.LocalModelCache, 0)
 
-	uncachedModelCount := 0
+	modelsThatWereCachedAlready := 0
+	for i, model := range apiSpec.CuratedModelResources {
+		if onlyLocalModels && model.S3Path {
+			continue
+		}
 
-	localModelCaches := make([]*spec.LocalModelCache, len(modelPaths))
-	for i, modelPath := range modelPaths {
-		var err error
-		modelCacheID, err := modelCacheID(modelPath, awsClient)
+		localModelCache, wasAlreadyCached, err = cacheLocalModel(model)
 		if err != nil {
 			if apiSpec.Predictor.ModelPath != nil {
-				return nil, errors.Wrap(err, apiSpec.Identify(), userconfig.PredictorKey, userconfig.ModelPathKey)
+				return errors.Wrap(err, apiSpec.Identify(), userconfig.PredictorKey, userconfig.ModelPathKey)
+			} else if apiSpec.Predictor.Models != nil && apiSpec.Predictor.Models.Dir != nil {
+				return errors.Wrap(err, apiSpec.Identify(), userconfig.PredictorKey, userconfig.ModelsKey, userconfig.ModelsDirKey, apiSpec.CuratedModelResources[i].Name, *apiSpec.Predictor.Models.Dir)
 			}
-			return nil, errors.Wrap(err, apiSpec.Identify(), userconfig.PredictorKey, userconfig.ModelsKey, apiSpec.Predictor.Models[i].Name, userconfig.ModelPathKey)
+			return errors.Wrap(err, apiSpec.Identify(), userconfig.PredictorKey, userconfig.ModelsKey, userconfig.ModelsPathsKey, apiSpec.CuratedModelResources[i].Name, userconfig.ModelPathKey)
 		}
-
-		localModelCache := spec.LocalModelCache{
-			ID:         modelCacheID,
-			HostPath:   filepath.Join(_modelCacheDir, modelCacheID),
-			TargetPath: apiSpec.Predictor.Models[i].Name,
+		if wasAlreadyCached {
+			modelsThatWereCachedAlready++
 		}
-
-		if !files.IsFile(filepath.Join(localModelCache.HostPath, "_SUCCESS")) {
-			err = cacheModel(modelPath, localModelCache, awsClient)
-			if err != nil {
-				if apiSpec.Predictor.ModelPath != nil {
-					return nil, errors.Wrap(err, apiSpec.Identify(), userconfig.PredictorKey, userconfig.ModelPathKey)
-				}
-				return nil, errors.Wrap(err, apiSpec.Identify(), userconfig.PredictorKey, userconfig.ModelsKey, apiSpec.Predictor.Models[i].Name, userconfig.ModelPathKey)
-			}
-			uncachedModelCount++
+		if len(model.Versions) == 0 {
+			localModelCache.TargetPath = filepath.Join(apiSpec.CuratedModelResources[i].Name, "1")
+		} else {
+			localModelCache.TargetPath = apiSpec.CuratedModelResources[i].Name
 		}
 
-		localModelCaches[i] = &localModelCache
+		localModelCaches = append(localModelCaches, localModelCache)
 	}
+	apiSpec.LocalModelCaches = localModelCaches
 
-	if uncachedModelCount > 0 {
+	if len(localModelCaches) > modelsThatWereCachedAlready && len(localModelCaches) > 0 {
```

Comment:
Not exactly. Even if you have multiple local models, it doesn't necessarily mean that the line break has to be printed. It only needs to be printed if a model got cached now. Otherwise, an extra print line would be printed when all requested models are already cached.